File: 01_02_wsmt/middleware/client-java/app/src/main/java/wsmt/mw/Library.java
package wsmt.mw;

import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.util.Date;
import java.util.List;
import java.util.concurrent.Callable;

import net.razorvine.pyro.PyroProxy;

@Command(name = "library", headerHeading = "Manage authors/books")
class Library implements Callable<Integer> {
  @Parameters(index = "0", description = "authors/books")
  private String entity;

  @Parameters(index = "1", description = "get/add/delete/update")
  private String operation;

  @Option(names = { "-n", "--name" }, description = "The name of the author for add/update")
  private String name = "";

  @Option(names = { "-t", "--title" }, description = "The title of the book for add/update")
  private String title = "";

  @Option(names = { "-q", "--query" }, description = "Substring to look for in author/book names/titles")
  private String query = "";

  @Option(names = { "-d", "--date" }, description = "The publish date or author birthday")
  private Date date;

  @Option(names = { "-id",
      "--id" }, description = "The id of the author/book for update/delete or the id of the author for book add")
  private int id;

  private PyroProxy proxy;

  public Library(PyroProxy proxy) {
    super();
    this.proxy = proxy;
  }

  @Override
  public Integer call() throws Exception {
    if ("books".equals(entity))
      switch (operation) {
        case "get":
          @SuppressWarnings("unchecked")
          List<Object> books = (List<Object>) proxy.call("get_books", query);
          for (Object book : books) {
            System.out.println(book);
          }
          break;
        case "add":
          proxy.call("add_book", title, id, date);
          break;
        case "delete":
          proxy.call("delete_book", id);
          break;
        case "update":
          proxy.call("update_book", id, title);
          break;
        default:
          return 4;
      }
    else if ("authors".equals(entity)) {
      switch (operation) {
        case "get":
          @SuppressWarnings("unchecked")
          List<Object> authors = (List<Object>) proxy.call("get_authors", query);
          for (Object author : authors) {
            System.out.println(author);
          }
          break;
        case "add":
          proxy.call("add_author", name, date);
          break;
        case "delete":
          proxy.call("delete_author", id);
          break;
        case "update":
          proxy.call("update_author", id, name);
          break;
        default:
          return 3;
      }
    } else {
      return 2;
    }
    return 0;
  }
}

File: 01_02_wsmt/middleware/server/library.py
from datetime import datetime
from typing import List
from models import Book, Person


class LibraryService:
    def add_book(self, title: str, author_id: int, published_at: datetime):
        author = Person.get_by_id(author_id)
        return Book.create(title=title, author=author, published_at=published_at)

    def get_books(self, q: str):
        query = Book.select().join(Person)
        if q != '':
            query = query.where(Book.title.contains(q))

        books: List[Book] = []
        for b in query:
            books.append(b)

        return books

    def update_book(self, id: int, title: str):
        return Book.get_by_id(id).update(title=title)

    def delete_book(self, id: int):
        return Book.delete_by_id(id)

    def add_author(self, name: str, birthday: datetime):
        return Person.create(name=name, birthday=birthday)

    def get_authors(self, q: str):
        query = Person.select()
        if q != '':
            query = query.where(Person.name.contains(q))

        authors: List[Person] = []
        for p in query:
            authors.append(p)

        return authors

    def update_author(self, id: int, name: str):
        return Person.get_by_id(id).update(name=name)

    def delete_author(self, id: int):
        return Person.delete_by_id(id)
